mbf_abstract_nav/src/abstract_planner_execution.cpp

```diff
@@ -261,47 +256,48 @@ void AbstractPlannerExecution::run()
         current_tolerance = tolerance_;
         ROS_INFO_STREAM("A new goal pose is available. Planning with the new goal pose and the tolerance: "
                         << current_tolerance);
-        exceeded = false;
         geometry_msgs::Point g = goal_.pose.position;
         ROS_INFO_STREAM("New goal pose: (" << g.x << ", " << g.y << ", " << g.z << ")");
       }
 
-      make_plan = !(success || exceeded) || has_new_start_ || has_new_goal_;
-
       // unlock goal
       goal_start_mtx_.unlock();
-      setState(PLANNING, false);
-      if (make_plan)
+      if (cancel_)
+      {
+        ROS_INFO_STREAM("The global planner has been canceled!");
+        planning_ = false;
+        setState(CANCELED, true);
+      }
+      else
       {
+        setState(PLANNING, false);
+
         outcome_ = makePlan(current_start, current_goal, current_tolerance, plan, cost, message_);
-        success = outcome_ < 10;
+        bool success = outcome_ < 10;
+
+        // assume we are done (will change this below if we need to retry)
+        planning_ = false;
 
         boost::lock_guard<boost::mutex> guard(configuration_mutex_);
 
         if (cancel_ && !isPatienceExceeded())
         {
           ROS_INFO_STREAM("The planner \"" << name_ << "\" has been canceled!"); // but not due to patience exceeded
-          planning_ = false;
           setState(CANCELED, true);
         }
         else if (success)
         {
           ROS_DEBUG_STREAM("Successfully found a plan.");
-          exceeded = false;
-          planning_ = false;
 
-          plan_mtx_.lock();
+          boost::lock_guard<boost::mutex> plan_mtx_guard(plan_mtx_);
           plan_ = plan;
           cost_ = cost;
           last_valid_plan_time_ = ros::Time::now();
-          plan_mtx_.unlock();
           setState(FOUND_PLAN, true);
         }
-        else if (max_retries_ >= 0 && ++retries > max_retries_)
+        else if (max_retries_ > 0 && ++retries > max_retries_)
```

Comment:
mind here the strict greater. With the >= case, we can never enter NO_PLAN_FOUND case